cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
+				_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+			}
 
-		WorkspaceProvider workspaceProvider = bladeCLI.getWorkspaceProvider(baseDir);
+			String absolutePattern = patternPrefix + pattern;
 
-		File workspaceDir = workspaceProvider.getWorkspaceDir(bladeCLI);
+			_addPathMatcher(pathMatchers, fileSystem, absolutePattern);
+		}
+
+		return pathMatchers;
+	}
 
-		Set<String> projectPaths = projectOutputFiles.keySet();
+	private Map<String, Path> _getProjectPaths(
+			final Path watchPath, List<String> projectPaths, List<String> ignorePaths)
+		throws Exception {
 
-		try (final GogoShellClient client = new GogoShellClient()) {
-			for (String projectPath : projectPaths) {
-				File projectDir = new File(workspaceDir, projectPath.replaceAll(":", File.separator));
+		final Map<String, Path> foundProjectPaths = new HashMap<>();
 
-				File watchOutputFile = new File(projectDir, "build/installedBundleId");
+		FileSystem fileSystem = FileSystems.getDefault();
 
-				if (watchOutputFile.exists()) {
-					try {
-						String installedBundleId = new String(Files.readAllBytes(watchOutputFile.toPath()));
+		List<PathMatcher> ignorePathMatchers = ignorePaths.stream(
+		).map(
+			ignorePath -> fileSystem.getPathMatcher("glob:" + ignorePath)
+		).collect(
+			Collectors.toList()
+		);
 
-						String response = client.send("uninstall " + installedBundleId);
+		Files.walkFileTree(
+			watchPath,
+			new SimpleFileVisitor<Path>() {
 
-						bladeCLI.out(response);
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					boolean shouldIgnorePath = ignorePathMatchers.stream(
+					).filter(
+						pathMatcher -> pathMatcher.matches(path)
+					).findFirst(
+					).isPresent();
+
+					if (shouldIgnorePath) {
+						return FileVisitResult.SKIP_SUBTREE;
 					}
-					catch (IOException ioe) {
+
+					try (Stream<Path> files = Files.list(path)) {
+						boolean projectPathFound = files.map(
+							p -> p.getFileName()
+						).filter(
+							p -> !ignorePathMatchers.stream(
+							).filter(
+								pathMatcher -> pathMatcher.matches(path.resolve(p))
+							).findFirst(
+							).isPresent()
+						).filter(
+							p -> projectPaths.stream(
+							).filter(
+								pp -> Objects.equals(pp, p.toString())
+							).findFirst(
+							).isPresent()
+						).findFirst(
+						).isPresent();
+
+						if (projectPathFound) {
+							foundProjectPaths.put(_getGradlePath(path, watchPath), path);
+
+							return FileVisitResult.SKIP_SUBTREE;
+						}
 					}
+
+					return FileVisitResult.CONTINUE;
 				}
-			}
-		}
-		catch (IOException ioe) {
-		}
+
+			});
+
+		return foundProjectPaths;
 	}
 
-	private void _watch(Path watchPath, Map<String, Set<File>> projectOutputFiles) throws InterruptedException {
+	private void _registerDirectory(WatchService watcher, Map<WatchKey, Path> keys, Path dir) throws IOException {
```

Comment:
Could this code be called on the same `dir` multiple times, and is that an issue?

Reply:
I'm not for sure actually @david-truong  do you know?